### ISSUE TITLE
Getting wrong results when testing cnn_mnist

### ISSUE THREAD

**Issue by rzadiki** [vlfeat/matconvnet#193]
Hi,

I'm trying to run the cnn_mnist example by training the net and then calling vl_simplenn() with couple of input images.
After the training phase, I modified the last layer to 'softmax' (instead of 'softmaxloss') and ran two different images taken from the training DB.
Foe some reason I get the exact same (wrong) scores for every tested image (please see example below).
Could you please help me understand what am I doing wrong?
Thanks!

>> [net_bn, info_bn] = cnn_mnist('expDir', 'data/mnist-bnorm', 'useBnorm', true);
>> net_bn1 = net_bn;
>> net_bn1.layers{1,11}.type = 'softmax';

>> DB = load('data\mnist-bnorm\imdb.mat')
>> im_preprocessed = DB.images.data(:,:,:,4);
>> res = vl_simplenn(net_bn1, im_preprocessed );
>> scores = squeeze(gather(res(end).x)) ;
>> scores'

ans =

    0.0984    0.0987    0.0997    0.1015    0.0988    0.1006    0.0977    0.1005    0.1030    0.1011

>> im_preprocessed = DB.images.data(:,:,:,5);
>> imagesc(im_preprocessed)
>> res = vl_simplenn(net_bn1, im_preprocessed );
>> scores = squeeze(gather(res(end).x)) ;
>> scores'

ans =

    0.0984    0.0987    0.0997    0.1015    0.0988    0.1006    0.0977    0.1005    0.1030    0.1011

**Comment by vedaldi** [comment]
Hi, I think the problem is in the bnorm layers. After training with bnorm layers, these should be removed as they do not work properly without at least a mini batch of data (as used in training). This is slightly annoying as it requires estimating the mean and covariance of each bnorm layer input  in the network (averaging over a sufficiently large number of mini batches), and then incorporating these values in the filter bank following bnorm, reproducing the bnorm effect, but for fixed mean and covariance.

In a future version we will try to make this step automatic.

**Comment by AruniRC** [comment]
Hi, just to clarify: you suggested that for now, we can run a network trained using bnorm layers by simply removing those layers and running it. 

would this affect the accuracy of the net at testing time a lot? -- the layer weights were trained to expect normalized inputs, and now those bnorm layers are removed. so the distribution of the inputs for which the network was trained is different from the inputs at test time.

**Comment by vedaldi** [comment]
Hi, the correct solution is to estimate such variances and biases with a small script over a bunch of batches and incorporate that (fixed) scaling/bias factors into the filters. We are thinking how this could be automatised in order to be transparent to a user, but at the moment there is no automatic provision to do so.

**Comment by AruniRC** [comment]
Thank you for the clarification.

**Comment by dbbert** [comment]
Is there any update on this issue? @AruniRC, do you have a script to determine the scaling/bias factors  of each layer over the entire training set?
